### ISSUE TITLE
Data export of native query not working if columns are reorganized without query permissions

### ISSUE THREAD

**Issue by ajcoll5** [metabase/metabase#16836]
**Describe the bug**
If the user doesn't have query permissions and reorganizes the columns of a saved native SQL query, the exports will contain a "missing-required-permissions" error. User is able to export fine if they do not change column order.

**Logs**
```js
{"database_id":2,"started_at":"2021-06-30T16:04:57.42403Z","error_type":"missing-required-permissions","json_query":{"type":"native","middleware":{"skip-results-metadata?":true,"format-rows?":false},"native":{"template-tags":{"AgingDate":{"id":"8205f989-c3c6-082b-b729-74d01d753d17","name":"AgingDate","display-name":"Aging Date","type":"date","required":false},"StartVendor":{"id":"2f0ff4db-35b4-ff3f-cc99-c60f71102a4f","name":"StartVendor","display-name":"Vendor Start #","type":"text"},"EndVendor":{"id":"8824e03b-a908-78c8-7b3b-c3ca85e45be2","name":"EndVendor","display-name":"Vendor End #","type":"text"}},"query":"<snip for privacy>"},"database":2,"async?":true,"cache-ttl":13},"status":"failed","class":"class clojure.lang.ExceptionInfo","stacktrace":["--> query_processor.middleware.permissions$perms_exception.invokeStatic(permissions.clj:27)","query_processor.middleware.permissions$perms_exception.doInvoke(permissions.clj:22)","query_processor.middleware.permissions$perms_exception.invokeStatic(permissions.clj:24)","query_processor.middleware.permissions$perms_exception.invoke(permissions.clj:22)","query_processor.middleware.permissions$fn__46594$check_ad_hoc_query_perms__46599$fn__46603.invoke(permissions.clj:60)","query_processor.middleware.permissions$fn__46594$check_ad_hoc_query_perms__46599.invoke(permissions.clj:49)","query_processor.middleware.permissions$fn__46631$check_query_permissions_STAR___46636$fn__46637.invoke(permissions.clj:71)","query_processor.middleware.permissions$fn__46631$check_query_permissions_STAR___46636.invoke(permissions.clj:64)","query_processor.middleware.permissions$check_query_permissions$fn__46650.invoke(permissions.clj:80)","query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__48768.invoke(pre_alias_aggregations.clj:40)","query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__46851.invoke(cumulative_aggregations.clj:60)","query_processor.middleware.resolve_joined_fields$resolve_joined_fields$fn__49067.invoke(resolve_joined_fields.clj:102)","query_processor.middleware.resolve_joins$resolve_joins$fn__49380.invoke(resolve_joins.clj:171)","query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__44925.invoke(add_implicit_joins.clj:190)","query_processor.middleware.large_int_id$convert_id_to_string$fn__47614.invoke(large_int_id.clj:59)","query_processor.middleware.format_rows$format_rows$fn__47595.invoke(format_rows.clj:74)","query_processor.middleware.add_default_temporal_unit$add_default_temporal_unit$fn__44219.invoke(add_default_temporal_unit.clj:23)","query_processor.middleware.desugar$desugar$fn__46917.invoke(desugar.clj:21)","query_processor.middleware.binning$update_binning_strategy$fn__45736.invoke(binning.clj:227)","query_processor.middleware.resolve_fields$resolve_fields$fn__46453.invoke(resolve_fields.clj:34)","query_processor.middleware.add_dimension_projections$add_remapping$fn__44574.invoke(add_dimension_projections.clj:312)","query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__44803.invoke(add_implicit_clauses.clj:147)","query_processor.middleware.upgrade_field_literals$upgrade_field_literals$fn__49789.invoke(upgrade_field_literals.clj:40)","query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__45088.invoke(add_source_metadata.clj:123)","query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__48942.invoke(reconcile_breakout_and_order_by_bucketing.clj:100)","query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__45296.invoke(auto_bucket_datetimes.clj:147)","query_processor.middleware.resolve_source_table$resolve_source_tables$fn__46500.invoke(resolve_source_table.clj:45)","query_processor.middleware.parameters$substitute_parameters$fn__48750.invoke(parameters.clj:111)","query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__46552.invoke(resolve_referenced.clj:79)","query_processor.middleware.expand_macros$expand_macros$fn__47301.invoke(expand_macros.clj:184)","query_processor.middleware.add_timezone_info$add_timezone_info$fn__45097.invoke(add_timezone_info.clj:15)","query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__49742.invoke(splice_params_in_response.clj:32)","query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__48953$fn__48957.invoke(resolve_database_and_driver.clj:31)","driver$do_with_driver.invokeStatic(driver.clj:60)","driver$do_with_driver.invoke(driver.clj:56)","query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__48953.invoke(resolve_database_and_driver.clj:25)","query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__47541.invoke(fetch_source_query.clj:274)","query_processor.middleware.store$initialize_store$fn__49751$fn__49752.invoke(store.clj:11)","query_processor.store$do_with_store.invokeStatic(store.clj:44)","query_processor.store$do_with_store.invoke(store.clj:38)","query_processor.middleware.store$initialize_store$fn__49751.invoke(store.clj:10)","query_processor.middleware.validate$validate_query$fn__49796.invoke(validate.clj:10)","query_processor.middleware.normalize_query$normalize$fn__47677.invoke(normalize_query.clj:22)","query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__44943.invoke(add_rows_truncated.clj:35)","query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__49727.invoke(results_metadata.clj:143)","query_processor.middleware.constraints$add_default_userland_constraints$fn__46794.invoke(constraints.clj:42)","query_processor.middleware.process_userland_query$process_userland_query$fn__48839.invoke(process_userland_query.clj:135)","query_processor.middleware.catch_exceptions$catch_exceptions$fn__46734.invoke(catch_exceptions.clj:173)","query_processor.reducible$async_qp$qp_STAR___38051$thunk__38052.invoke(reducible.clj:103)","query_processor.reducible$async_qp$qp_STAR___38051$fn__38054.invoke(reducible.clj:108)"],"context":"json-download","error":"You do not have permissions to run this query.","row_count":0,"running_time":0,"ex-data":{"type":"missing-required-permissions","required-permissions":["/db/2/native/"],"actual-permissions":["/collection/7/read/","/db/2/schema/dbo/table/709/","/db/2/schema/dbo/table/949/","/collection/10/read/","/collection/4/","/collection/51/","/collection/51/read/","/collection/5/read/","/collection/14/read/","/collection/12/read/","/collection/2/read/","/db/2/schema/dbo/table/1350/","/collection/6/"],"permissions-error?":true},"data":{"rows":[],"cols":[]}}`
```
**To Reproduce**
1. Go to a saved native SQL query on a user without SQL query permissions.
2. Click and drag any column to reorganize it.
3. Click on "Download full results" and select any format.
4. See error in export.

**Expected behavior**
The export should happen without issue - ideally with the columns in the order that is displayed, but at worst the order of the original query.

**Information about your Metabase Installation:**

- Your browser and the version: Firefox 90.0b12
- Your operating system: Windows 10 21H1
- Your databases: SQL Server
- Metabase version: 0.39.4
- Metabase hosting environment: Docker (official image)
- Metabase internal database: PostgreDB

**Severity**
Low - can be worked around by either not reordering the columns or clicking back the "original" question before export.

**Comment by flamber** [comment]
Before 0.38.3 (#15424) you wouldn't even be able to reorganize columns (or any other changes) without seeing a permission error.
When trying to export the user gets the full page error `You don't have permissions to do that.`

**Comment by flamber** [comment]
@pldeepesh You are seeing this issue: #18433

**Comment by ajcoll5** [comment]
@pldeepesh - I don't believe I had tried applying filters and exporting before, only moving/resizing columns. We don't use SQL questions much as I try to keep most of our questions as standard Metabase questions to avoid the limited end-user functionality of SQL questions. For what it's worth, on 0.42.3 I can change variables and export fine with a read-only user on a SQL question that was last edited 9 months ago.

I've noticed with the latest version a read-only user can no longer move columns on SQL questions - making this bug report somewhat moot. A resize will still trigger the bug by making the question dirty, but from pull requests it looks like resizing will be removed soon too.
